src/messaging/ReliableMessageManager.cpp

```diff
@@ -295,66 +330,76 @@ void ReliableMessageManager::Timeout(System::Layer * aSystemLayer, void * aAppSt
  *
  *  @param[in]    rc        A pointer to the ExchangeContext object.
  *
- *  @param[in]    msgBuf    A pointer to the message buffer holding the CHIP message to be retransmitted.
+ *  @param[in]    header    A reference to the Exchange Header for the CHIP message to be retransmitted.
  *
- *  @param[in]    messageId The message identifier of the stored CHIP message.
+ *  @param[in]    nodeId    A peer Node ID of the CHIP message to be retransmitted.
  *
  *  @param[out]   rEntry    A pointer to a pointer of a retransmission table entry added into the table.
  *
  *  @retval  #CHIP_ERROR_RETRANS_TABLE_FULL If there is no empty slot left in the table for addition.
  *  @retval  #CHIP_NO_ERROR On success.
  *
  */
-CHIP_ERROR ReliableMessageManager::AddToRetransTable(ReliableMessageContext * rc, System::PacketBuffer * msgBuf, uint32_t messageId,
-                                                     uint16_t msgSendFlags, RetransTableEntry ** rEntry)
+CHIP_ERROR ReliableMessageManager::AddToRetransTable(ReliableMessageContext * rc, PayloadHeader & header, NodeId nodeId,
+                                                     RetransTableEntry ** rEntry)
 {
     bool added     = false;
     CHIP_ERROR err = CHIP_NO_ERROR;
 
+    VerifyOrDie(rc != nullptr && rc->mExchange != nullptr);
+
     for (int i = 0; i < CHIP_CONFIG_RMP_RETRANS_TABLE_SIZE; i++)
     {
         // Check the exchContext pointer for finding an empty slot in Table
-        if (!RetransTable[i].rc)
+        if (!mRetransTable[i].rc)
         {
             // Expire any virtual ticks that have expired so all wakeup sources reflect the current time
             ExpireTicks();
 
-            RetransTable[i].rc                  = rc;
-            RetransTable[i].msgId               = messageId;
-            RetransTable[i].msgBuf              = msgBuf;
-            RetransTable[i].msgSendFlags        = msgSendFlags;
-            RetransTable[i].sendCount           = 0;
-            RetransTable[i].nextRetransTimeTick = static_cast<uint16_t>(
-                rc->GetCurrentRetransmitTimeoutTick() + GetTickCounterFromTimeDelta(System::Timer::GetCurrentEpoch()));
+            mRetransTable[i].rc            = rc;
+            mRetransTable[i].payloadHeader = header;
+            mRetransTable[i].peerNodeId    = nodeId;
+            mRetransTable[i].sendCount     = 0;
+            mRetransTable[i].retainedBuf   = System::EncryptedPacketBufferHandle();
+
+            *rEntry = &mRetransTable[i];
 
-            *rEntry = &RetransTable[i];
             // Increment the reference count
-            rc->Retain();
+            rc->mExchange->Retain();
```

Comment:
Yes, the lifetime of rc is managed by its exchange, but the lifetime of exchange is managed by both protocol and ReliableMessageManager.

The protocol layer is the only entity that knows the sequence of messages that are to be part of a given exchange - the ReliableMessageManager doesn’t know this, and as such, the protocol effectively “owns” the exchange once it’s been vended by the ExchangeMgr. There is a transfer of ownership however, when the protocol layer decides to close the exchange, at which point the ReliableMessageManager handles the remaining work to be done on that exchange before truly cleaning it up (like closing the connection, or waiting for a CRMP ack to a message).

When a rEntry is added to the retransTable of ReliableMessageManager, its exchange need to be managed by ReliableMessageManager, so we need to retain the exchange at this point.

Reply:
I understand why we need the retain call, yes.

My point is that the thing we want to keep alive is the `rc`, because that's what we have a pointer to.  The fact that calling `rc->mExchange->Retain()` will keep both `rc->mExchange` _and_ `rc` alive is highly non-obvious.  Whereas if we had `rc->Retain()` it would be clear what we are doing here: keeping alive the thing that we are actually going to reference later.  This means it will keep alive whatever it needs to function properly, including the exchange.

The actual implementation of `ReliableMessageContext::Retain` would be:
```
  void Retain() { mExchange->Retain(); }
```
and similar for releasing/freeing it.  But it will be much clearer at the callsite that "yes, we are keeping alive the thing we have a pointer to".

Reply:
Acked